subsys/bluetooth/audio/has.c

```diff
@@ -34,7 +56,362 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
+	.disconnected = disconnected,
+};
+
+typedef uint8_t (*preset_func_t)(const struct bt_has_preset *preset, void *user_data);
+
+static void preset_foreach(uint8_t start_index, uint8_t end_index, preset_func_t func,
+			   void *user_data)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+		const struct bt_has_preset *preset = &has_preset_list[i];
+
+		/* Skip ahead if out of range */
+		if (preset->index < start_index  || preset->index > end_index) {
+			continue;
+		}
+
+		if (func(preset, user_data) == BT_HAS_PRESET_ITER_STOP) {
+			return;
+		}
+	}
+}
+
+static uint8_t preset_found(const struct bt_has_preset *preset, void *user_data)
+{
+	const struct bt_has_preset **found = user_data;
+
+	*found = preset;
+
+	return BT_HAS_PRESET_ITER_STOP;
+}
+
+static struct bt_has_preset *preset_alloc(uint8_t index, enum bt_has_properties properties,
+					  const char *name)
+{
+	struct bt_has_preset *preset = NULL;
+
+	preset_foreach(BT_HAS_PRESET_INDEX_NONE, BT_HAS_PRESET_INDEX_NONE, preset_found, &preset);
+
+	if (preset != NULL) {
+		preset->index = index;
+		preset->properties = properties;
+#if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
+		utf8_lcpy(preset->name, name, ARRAY_SIZE(preset->name));
+		preset->name[strlen(name)] = '\0';
+#else
+		preset->name = name;
+#endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
+	}
+
+	return preset;
+}
+
+static void preset_free(struct bt_has_preset *preset)
+{
+	preset->index = BT_HAS_PRESET_INDEX_NONE;
+}
+
+static void control_point_ntf_complete(struct bt_conn *conn, void *user_data)
+{
+	struct has_client *client = client_get(conn);
+
+	BT_DBG("conn %p\n", (void *)conn);
+
+	/* Resubmit if needed */
+	if (read_presets_req_is_pending(client)) {
+		k_work_submit(&client->control_point_work);
+	}
+}
+
+static void control_point_ind_complete(struct bt_conn *conn,
+				       struct bt_gatt_indicate_params *params,
+				       uint8_t err)
+{
+	if (err) {
+		/* TODO: Handle error somehow */
+		BT_ERR("conn %p err 0x%02x\n", (void *)conn, err);
+	}
+
+	control_point_ntf_complete(conn, NULL);
+}
+
+static int control_point_send(struct has_client *client, struct net_buf_simple *buf)
+{
+#if defined(CONFIG_BT_EATT)
+	/* TODO: Check whether EATT is connected */
+	if (bt_gatt_is_subscribed(client->conn, &has_svc.attrs[4], BT_GATT_CCC_NOTIFY)) {
+		client->params.ntf.attr = &has_svc.attrs[4];
+		client->params.ntf.func = control_point_ntf_complete;
+		client->params.ntf.data = buf->data;
+		client->params.ntf.len = buf->len;
+
+		return bt_gatt_notify_cb(client->conn, &client->params.ntf);
+	}
+#endif /* CONFIG_BT_EATT */
+
+	if (bt_gatt_is_subscribed(client->conn, &has_svc.attrs[4], BT_GATT_CCC_INDICATE)) {
+		client->params.ind.attr = &has_svc.attrs[4];
+		client->params.ind.func = control_point_ind_complete;
+		client->params.ind.destroy = NULL;
+		client->params.ind.data = buf->data;
+		client->params.ind.len = buf->len;
+
+		return bt_gatt_indicate(client->conn, &client->params.ind);
+	}
```

Comment:
```suggestion
	const struct bt_gatt_attr *attr = &has_svc.attrs[4];
	
	/* TODO: Check whether EATT is connected */
	if (IS_ENABLED(CONFIG_BT_EATT) &&
	    bt_eatt_count(client->conn) > 0 &&
	    bt_gatt_is_subscribed(client->conn, attr, BT_GATT_CCC_NOTIFY)) {
		client->params.ntf.attr = attr;
		client->params.ntf.func = control_point_ntf_complete;
		client->params.ntf.data = buf->data;
		client->params.ntf.len = buf->len;

		return bt_gatt_notify_cb(client->conn, &client->params.ntf);
	}

	if (bt_gatt_is_subscribed(client->conn, attr, BT_GATT_CCC_INDICATE)) {
		client->params.ind.attr = attr;
		client->params.ind.func = control_point_ind_complete;
		client->params.ind.destroy = NULL;
		client->params.ind.data = buf->data;
		client->params.ind.len = buf->len;

		return bt_gatt_indicate(client->conn, &client->params.ind);
	}
```

and then possible give `attr` a less generic name, e.g. the name of the characteristic

Reply:
You can also consider using `bt_gatt_notify_uuid` to avoid the specific index in the attributes

Reply:
Unfortunately I cannot use `IS_ENABLED` because I get compilation warning:
```
/home/mariuszskamra/zephyrproject/zephyr/subsys/bluetooth/audio/has.c:323:43: warning: implicit declaration of function ‘bt_eatt_count’ [-Wimplicit-function-declaration]
  323 |         if (IS_ENABLED(CONFIG_BT_EATT) && bt_eatt_count(client->conn) > 0 &&
```
moreover, the `ntf` parameters are conditional
```
#if defined(CONFIG_BT_EATT)
		struct bt_gatt_notify_params ntf;
#endif /* CONFIG_BT_EATT */
```

Reply:
Right, because `bt_eatt_count` is actually guarded by `CONFIG_BT_EATT` in the header file. A shame; that probably needs to be updated.